Sources/NIOSSL/SSLContext.swift

```diff
@@ -112,6 +112,119 @@ private func alpnCallback(ssl: OpaquePointer?,
+    let contextPSK = contextPSKIdentityCallback.key // From the context
+    let contextPSKIdentity = contextPSKIdentityCallback.identity // From the context
+    
+    // Equate both identities above to make sure both are the same (they should be as the context was copied into the incoming identity).
+    if contextPSKIdentity != incomingPSKIndentity {
+        // Warn about this here but do not fail.
+        print("PSK warning: client identity not what we expected (got \(Array(incomingPSKIndentity)) expected \(Array(contextPSKIdentity))")
+    }
+    if max_psk_len > UInt32.max || contextPSK.count > max_psk_len {
+        return 0
+    }
+    memcpy(incomingPSK, Array(contextPSK), contextPSK.count)
+    return UInt32(contextPSK.count)
+}
+
+/// PSK Callback for the client side context.
+private func clientPSKCallback(ssl: OpaquePointer?,
+                                 hint: UnsafePointer<CChar>?,
+                                 identity: UnsafeMutablePointer<CChar>?,
+                                 max_identity_len: UInt32,
+                                 psk: UnsafeMutablePointer<UInt8>?,
+                                 max_psk_len: UInt32) -> UInt32 {
+    
+    guard let ssl = ssl else { return 0 }
+    // Initial implementation only supports TLS 1.2 due API support exposed from BoringSSL.
+    // TODO (meaton) add TLS 1.3 support when available.
+    
+    // We want to take the SSL pointer and extract the parent Swift object.
+    let parentCtx = CNIOBoringSSL_SSL_get_SSL_CTX(ssl)!
+    let parentPtr = CNIOBoringSSLShims_SSL_CTX_get_app_data(parentCtx)!
+    let parentSwiftContext: NIOSSLContext = Unmanaged.fromOpaque(parentPtr).takeUnretainedValue()
+    
+    guard let clientCallback = parentSwiftContext.pskIdentityCallbackManager?.clientCallback,
+          let incomingIdentity = identity, // Incoming PSK identity. (Could be empty)
+          let incomingPSK = psk // Incoming PSK key.
+          else {
+        return 0
+    }
+    var incomingPSKIndentity = NIOSSLSecureBytes()
+    let strBasedIncomingIdentity = String(validatingUTF8: incomingIdentity) ?? ""
+
+    // If set, build out a server hint to pass into the client callback.
+    var serverHint = NIOSSLSecureBytes()
+    if let strServerHint = hint,
+       let derivedHint = String(validatingUTF8: strServerHint) {
+        serverHint.append(derivedHint.utf8)
+    }
+    
+    let contextPSKIdentityCallback = clientCallback(serverHint)
+    let contextPSK = contextPSKIdentityCallback.key // From the context
+    let contextPSKIdentity = contextPSKIdentityCallback.identity // From the context
+    
+    // If the server is sending the identity, store it first as incomingPSKIndentity and do not fail because of this.
+    if strBasedIncomingIdentity.count > 0 {
+        incomingPSKIndentity.append(strBasedIncomingIdentity.utf8)
+    } else if let sslIdentity = CNIOBoringSSL_SSL_get_psk_identity(ssl),
```

Comment:
This branch is also not relevant: the only identity that matters is the one the user returned.